src/NUnitFramework/framework/Constraints/Comparers/ComparisonState.cs

```diff
@@ -0,0 +1,36 @@
+using System.Collections.Generic;
+
+namespace NUnit.Framework.Constraints.Comparers
+{
+    internal struct ComparisonState
+    {
+        public struct Comparison
+        {
+            public object X { get; set; }
+            public object Y { get; set; }
+        }
+
+        /// <summary>
+        /// Flag indicating whether or not this is the top level comparison.
+        /// </summary>
+        public bool TopLevelComparison { get; set; }
+        public List<Comparison> Comparisons { get; set; }
+
+        public void RecordComparison(object x, object y)
+        {
+            Comparisons.Add(new Comparison()
+            {
+                X = x,
+                Y = y
+            });
```

Comment:
Even though `Comparison` is a struct, please make X and Y get-only properties and use `new Comparison(x, y)`. Guarantees of immutability reduce the cognitive load on the reviewers and future maintainers of the code.